typed-th-in-ghc-9/article.md

```diff
@@ -0,0 +1,205 @@
++    -- examineCode will turn 'Code' into 'Q'
+-    Nothing -> [|| Production ||]  -- We default to Production if the flag is unset
++    Nothing -> examineCode [|| Production ||]  -- We default to Production if the flag is unset
+    Just level -> case level of
+_      "DEVELOPMENT" -> [|| Development ||]
+_      "PRODUCTION" -> [|| Production ||]
++      "DEVELOPMENT" -> examineCode [|| Development ||]
++      "PRODUCTION" -> examineCode [|| Production ||]
+      other -> fail $ "Unrecognized LOG_LEVEL flag: " <> other
+```
+
+The errors in the previous section happened because the type returned by the quotations is no longer a `Q`, but `Code` instead. The solution is to use `examineCode` to turn a `Code m a` into a `m (TExp a)`, and then `liftCode` to do the opposite operation: turning a `m (TExp a)` into a `Code m a`.
+
+```hs
+liftCode    :: forall (r :: RuntimeRep) (a :: TYPE r) m. m (TExp a) -> Code m a
+examineCode :: forall (r :: RuntimeRep) (a :: TYPE r) m. Code m a -> m (TExp a)
+```
+
+And that's it! If you want the code to still work on GHC 8, however, make sure to read the next section.
+
+## Porting old TTH code with backward compatibility
+
+For this section, you may choose to use either GHC 8 or GHC 9. We will use the [`th-compat`](https://hackage.haskell.org/package/th-compat) package, besides the familiar [`template-haskell`](https://hackage.haskell.org/package/template-haskell) package, so make sure you load them.
+
+The pattern is pretty similar to the workflow with `Code`, albeit with `Splice` now.
+
+Make sure to import the appropriate definitions first:
+
+```hs
+import Language.Haskell.TH.Syntax (Q)
+import Language.Haskell.TH.Syntax.Compat (Splice, examineSplice, liftSplice)
+```
+
+And now just replace `Code` with `Splice`:
+
+```diff
+-logLevelFromFlag :: Code Q LogLevel
+-logLevelFromFlag = liftCode $ do  -- liftSplice will turn 'Q' into 'Code'
++logLevelFromFlag :: Splice Q LogLevel
++logLevelFromFlag = liftSplice $ do  -- liftSplice will turn 'Q' into 'Splice'
+  flag <- liftIO $ lookupEnv "LOG_LEVEL"
+  case flag of
+-    -- examineCode will turn 'Code' into 'Q'
+-    Nothing -> examineCode [|| Production ||]  -- We default to Production if the flag is unset
++    -- examineSplice will turn 'Splice' into 'Q'
++    Nothing -> examineSplice [|| Production ||]  -- We default to Production if the flag is unset
+    Just level -> case level of
+-      "DEVELOPMENT" -> examineCode [|| Development ||]
+-      "PRODUCTION" -> examineCode [|| Production ||]
++      "DEVELOPMENT" -> examineSplice [|| Development ||]
++      "PRODUCTION" -> examineSplice [|| Production ||]
+      other -> fail $ "Unrecognized LOG_LEVEL flag: " <> other
+```
+
+And that's it! `Splice m a` is defined in `th-compat` as `m (TExp a)` in GHC 8 and `Code m a` in GHC 9, and this is why this code works. In addition, functions like `liftSplice` and `examineSplice` will either be defined `liftCode` and `examineCode` (in GHC 9) or `id` (in GHC 8).
```

Comment:
My usual nitpick: it is not about GHC 9 vs. GHC 8 – it is about GHC >=9 vs. GHC < 9.

Reply:
Changed it.